File: C/0016ThreeSumClosest.c
#include <stdio.h>
#include <limits.h>
#include <stdlib.h>

int compare(const void *a, const void *b){
    int x = *(const int *)a;
    int y = *(const int *)b;

    if(x < y){
        return -1;
    }
    if(x > y){
        return 1;
    }

    return 0;
}

int threeSumClosest(int* nums, int numsSize, int target){
    qsort(nums, numsSize, sizeof(int), compare);

    int bestDiff = INT_MAX;
    int bestSum = 0;

    for(int i = 0; i < numsSize - 2; i++){
        int left = i + 1;
        int right = numsSize - 1;

        while(left < right){
            int currSum = nums[i] + nums[left] + nums[right];
            int currDiff = abs(currSum - target);

            if(currDiff < bestDiff){
                bestDiff = currDiff;
                bestSum = currSum;
            }

            if(currSum < target){
                left++;
            } else if(currSum > target){
                right--;
            } else{
                return currSum;
            }
        }
    }

    return bestSum;
}

int main(){

    int nums[] = {0, 0, 0};
    int numsSize = sizeof(nums) / sizeof(nums[0]);
    int target = 1;
    printf("%d", threeSumClosest(nums, numsSize, target));

    return 0;
}

File: Python/0016ThreeSumClosest.py
def threeSumClosest(nums : list[int], target : int) -> int:
    n = len(nums)
    nums.sort()

    bestDiff = float('inf')
    bestSum = 0

    for i in range(n - 2):
        left = i + 1
        right = n - 1

        while left < right:
            currSum = nums[i] + nums[left] + nums[right]
            diff = abs(currSum - target)
            
            if diff < bestDiff:
                bestDiff = diff
                bestSum = currSum
            
            if currSum < target:
                left += 1
            elif currSum > target:
                right -= 1
            else:
                return currSum
        
    return bestSum

nums = [0, 0, 0]
target = 1
print(threeSumClosest(nums, target))